### PR TITLE
[SYCL-MLIR][NFCI]: Introduce class CodeGenTypes, move code that deal with clang->MLIR types into it.

### PR DESCRIPTION
**Pull request by etiotto**
Introduce class CodeGenTypes, move code that deal with clang->MLIR types into it.

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CodeGenTypes.h [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+namespace clang {
+class ASTContext;
+template <typename> class CanQual;
+class CodeGenOptions;
+class FunctionDecl;
+class QualType;
+class RecordType;
+class Type;
+typedef CanQual<Type> CanQualType;
+class GlobalDecl;
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.h:30]:
```suggestion
class ASTContext;
class CodeGenOptions;
class QualType;
class RecordType;
class GlobalDecl;
```

File: polygeist/tools/cgeist/Lib/CodeGenTypes.h [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+namespace CodeGen {
+class ABIInfo;
+class CGCXXABI;
+class CGFunctionInfo;
+class CodeGenModule;
+} // namespace CodeGen
+} // namespace clang
+
+namespace mlir {
+class FunctionType;
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.h:41]:
```suggestion
```

File: polygeist/tools/cgeist/Lib/CodeGenTypes.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  const clang::CodeGenOptions &getCodeGenOpts() const;
+
+  // TODO: Possibly create a SYCLTypeCache
+  mlir::Type getMLIRType(clang::QualType QT, bool *ImplicitRef = nullptr,
+                         bool AllowMerge = true);
+
+  mlir::Type getPointerOrMemRefType(mlir::Type Ty, bool IsAlloca = false);
+
+  const clang::CodeGen::CGFunctionInfo &
+  arrangeGlobalDeclaration(clang::GlobalDecl GD);
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.h:82]:
```suggestion
```

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+                                 CGM.getDataLayout().getAllocaAddrSpace());
+
+  return LLVM::LLVMPointerType::get(Ty,
+                                    CGM.getDataLayout().getAllocaAddrSpace());
+}
+
+const clang::CodeGen::CGFunctionInfo &
+CodeGenTypes::arrangeGlobalDeclaration(clang::GlobalDecl GD) {
+  return CGM.getTypes().arrangeGlobalDeclaration(GD);
+}
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:639]:
```suggestion
```

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+      auto ty = getMLIRType(f->getType(), &subRef, /*allowMerge*/ false);
+      assert(!subRef);
+      innerLLVM |= ty.isa<LLVM::LLVMPointerType, LLVM::LLVMStructType,
+                          LLVM::LLVMArrayType>();
+      innerSYCL |=
+          ty.isa<mlir::sycl::IDType, mlir::sycl::AccessorType,
+                 mlir::sycl::NdRangeType, mlir::sycl::RangeType,
+                 mlir::sycl::AccessorImplDeviceType, mlir::sycl::ArrayType,
+                 mlir::sycl::ItemType, mlir::sycl::ItemBaseType,
+                 mlir::sycl::NdItemType, mlir::sycl::GroupType>();
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:400]:
```suggestion
      innerSYCL |= isSYCLType(ty);
```

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+      // When generating the sycl_halide_kernel, If a struct type contains
+      // SYCL types, that means that this is the functor, and we can't create
+      // a llvm pointer that contains custom aggregate types. We could create
+      // a sycl::Functor type, that will help us get rid of those conditions.
+      bool InnerSYCL = false;
+      if (auto ST = subType.dyn_cast<mlir::LLVM::LLVMStructType>()) {
+        InnerSYCL = any_of(ST.getBody(), [](auto Element) {
+          return mlir::sycl::isSYCLType(Element);
+        });
+      }
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:547]:
```suggestion
      if (auto ST = subType.dyn_cast<mlir::LLVM::LLVMStructType>())
        InnerSYCL |= any_of(ST.getBody(),
                            [](auto Element) { return isSYCLType(Element); });
```

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+}
+
+namespace mlirclang {
+namespace CodeGen {
+
+namespace {
+
+/// Encapsulates information about the way function arguments from
+/// CGFunctionInfo should be passed to actual LLVM IR function.
+class ClangToLLVMArgMapping {
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:90]: Can we please have `ClangToLLVMArgMapping` class added in the non-nfc pr?

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+// Note: cgeist does not allow flattening struct function parameters. Need to
+// revisit.
+constexpr bool AllowStructFlattening = false;
+
+// Note: cgeist does not allow returning a struct via the parameter list. Need
+// to revisit.
+constexpr bool AllowSRet = false;
+
+// Note: cgesit does not allow returning 'inalloca'. Need to revisit.
+constexpr bool AllowInAllocaRet = false;
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/CodeGenTypes.cc:60]: Can we add these flags in the non-nfc pr?

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 9]
```diff
@@ -3366,20 +3396,21 @@ MLIRASTConsumer::createMLIRFunction(const FunctionToEmit &FTE,
...
-  setMLIRFunctionAttributes(function, FTE, lnk);
+  setMLIRFunctionAttributes(function, FTE, ShouldEmit);
   setMLIRFunctionParmsAttributes(function, parmDescriptors);
   setMLIRFunctionResultAttributes(function, resDescriptors);
 
-  /// Inject the MLIR function created in either the device module or in the
-  /// host module, depending on the calling context.
+  /// Inject the MLIR function created in either the device
+  /// module or in the host module, depending on the calling
+  /// context.
```
Comment (whitneywhtsang) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3413]:
```suggestion
  /// Inject the MLIR function created in either the device module or in the
  /// host module, depending on the calling context.
```

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc (hunk 8 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CodeGenTypes.cc:60]: done

File: polygeist/tools/cgeist/Lib/CodeGenTypes.cc (hunk 7 above)
Comment (etiotto) [reply polygeist/tools/cgeist/Lib/CodeGenTypes.cc:90]: done

**Review by whitneywhtsang — APPROVED** [review]

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 10]
```diff
@@ -3266,6 +3276,7 @@ void MLIRASTConsumer::createMLIRParameterDescriptors(
     const clang::FunctionDecl &FD,
     SmallVectorImpl<CodeGenUtils::ParmDesc> &parmDescriptors) {
   assert(parmDescriptors.empty() && "Expecting 'parmDescriptors' to be empty");
+  llvm::dbgs() << "\n-- Entering createMLIRParameterDescriptors --\n";
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:3279]: This is displaying a message every time this function is called. Please, wrap around `LLVM_DEBUG`.

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Lib/clang-mlir.cc:3279]: Done in https://github.com/intel/llvm/pull/7149.

**Pull request merged**